### DISCUSSION TITLE
[cleaner] Option for distributed cleaning

### DISCUSSION THREAD

**Discussion by TrevorBenson** [sosreport/sos#3476] [Ideas]
Subject: Proposal for Distributed Cleaner in Light of Performance Issues

---

Hi everyone,

I've been reflecting on the cleaner's performance and open issues, and the idea of a distributed version caught my attention. I haven't delved deep into the obfuscation functions to assess their compatibility, and I'm aware of the ongoing feature request #3097 to enhance parallelism.

Considering potential obstacles, including in syncing parsers, a few thoughts:

1. **Potential Resource Discrepancy**
   - The collector system might have significantly fewer resources than the systems it collects from, impacting overall cleaning performance.
   - In clusters, the master node often requires only a fraction of resources compared to cluster nodes, leading to sub-optimal cleaning speed as this system has a high probability of being the collector.

2. **Performance Insights**
   - Feedback in #3097 highlights severe performance impact on single archives compared to the current cleaning process ([comment link](https://github.com/sosreport/sos/issues/3097#issuecomment-1369068321)).
   - Negative impact on single archives caused by concurrent cleaning could be avoided in distributed cleaning operations. At least if the current cleaner without more parallism was distributed, it would have the same (or better performance depending on the nodes resources) and the work would be divided among the entire cluster.

I'm not suggesting this be the default cleaner option, but exploring distributed cleaning could address scenarios where waiting for the cleaner on an under-powered system is impractical. Therefore I feel it's worth considering the requirements for a distributed approach which could also ensure consistent obfuscation across all archives and the produced mapping file.

Looking forward to your thoughts and insights.

**Comment by pmoravec** [comment]
Running cleaner concurrently on individual cluster nodes can gain significant performance gain (while implementing *just* this feature shall be easy, esp. comparing to concurrent cleaner inside one archive). But the cost is "incompatible" mapping where e.g. the same cluster node name is mapped once as `host1` and once as `host12`, since the mapping on individual cluster nodes is being built independently.

Still this feature has sense even when having the full feature "cleaner runnng concurrently over one archive" for the performance reasons - user shall have an option to that.

**Reply by TrevorBenson** [reply]
Yeah, when I was thinking of this I was contemplating obfuscation functions which took a hash/salt used to create the obfuscated data. Basically the collector would define it, pass it along, and then with each node using the same hash/salt result in the same transformation of `host1`.

I suspected the current functions did not support anything like this, but didn't review them yet. It was also just a first thought when considering the concept of a distributed obfuscation with a unified mapping file in the end. 

I suspect there are better, more well thought out, ideas among the group on how this might be accomplished.

**Reply by pmoravec** [reply]
Using a global salt sounds a great idea (if I interpret it right that `sos` generates a secret salt that encodes a hostname (or MAC address etc) to a hash to constitute `host<HASH>` uniform on any system). This approach can be beneficial also for concurrent cleaner on one system (thought Jake started preparations for another approach - but that should not be in conflict here..?).

The salt should be stored somewhere on the disk to re-use it on subsequent calls of the cleaner (currently we store the mappings "only" for the same purposes). And also storing it in a file can be the mechanism how collector tells individual cluster nodes what salt to use.. Sadly this will mean if one runs sosreport locally on a node (which will generate a first salt) and then run it from a cluster, the cluster's salt will overwrite the local one and mappings from those sosreports wont be compatible. But I expect this use case as rare.

**Comment by TurboTurtle** [comment]
I'm certainly interested in hearing more about a 'distributed' cleaner, but first let me recap where the performance discussion was heading as of ~6 months ago.

I was looking at transitioning the flow to use a set of processes (as opposed to threads) that would be responsible for reading files in parallel and finding items for obfuscation. This obfuscation would be handled by having the process add the to-be-obfuscated item to a `Queue()` in the main process, and the main process would solely be dedicated to retrieving and answering the items in this queue.

This should, in theory at least, allow our performance to now be limited by how fast the main thread can pull and respond to items in the queue, rather than how many files we can read concurrently.

This involves at minimum decoupling `parser`s from holding the obfuscation maps, and instead having them literally just as fancy abstractions for regex patterns. The `mapping`s would be what the main process keeps control over to ensure obfuscations are consistent between processes.

Depending on the process management design, this could either be "1 process per archive" - akin to what the idea is today (but what we don't actually achieve) or it could be "just throw a bunch of processors at this huge list of files". The latter makes for an easier implementation to support multi-process single archive cleaning, but may not be the most efficient design as we'd probably need to enumerate all the files we'll be scanning before starting the process to ensure we're actively using all our desired processes at any given time. This in turn would likely mean we'd need to (or at least want to) unpack every archive before beginning the obfuscation process, which isn't great either. There's a lot of options here in the overall design to talk about.

All that said, we're a ways off right now. I was not able to finish decoupling the parsers and mappings before I left Red Hat. Then there is the fact that the queue-based workflow, while "simple" in design, is not a trivial amount of work on its own and would need significant testing to make sure our obfuscations are consistent regardless of scale.

**Comment by pmoravec** [comment]
There might be an "elegant" alternative approach, though it has its cons as well.

The problem in cleaning concurrently is to maintain a sequence of IDs in `obfuscateduser1`-like strings, sequentially. What about replacing the numerical ID by a hash (of length, say, 6-8 chars)? The hash would be deterministically computed from the original string that we obfuscate. And it would allow to create the mapping absolutely concurrently / independently. Just a merge at the end would have to happen.

There are a few gotchas, however.

First, the length of the hash. It must be reasonably short to be human-understandable. Gladly a typical string to be obfuscated is relatively short length, that can be represented by relatively short hash. Very rule-of-thumb estimation is 6-8 chars, imho, to prevent hash collisions. Some better estimation is welcomed ("we obfuscate words usually up to X chars, so a hash of 6/7/8 chars would mean probability of Y% of a collision = that is too much / adequate / perfect").

Then, we would need adding some salt, private and semi-unique for the system that triggers the cleaner. Otherwise, one can run a vocabulary attack to generate list of such short hashes and get the obfuscated word from the given hash.

This salt would bring a new problem: running sosreports from system A with saltA would be in conflict of running sos collect from system B with saltB (that will collect sosreport also from system A) - since we need to have one salt for whole sos collect. But I think this type of problems are common to the current approach with sequential IDs as well..?

Last problem I am aware is incompatibility of this mapping with the current one. This new mapping would have to throw away the current existing `/etc/sos/cleaner/default_mapping` from current "sequential IDs" mapping - once for every such system. Not a big deal, but e.g. it will make difficult to correlate obfuscated strings from old mapping (`obfuscateduser12`) to the new one (`obfuscateduser-asdf123`).

**Comment by pmoravec** [comment]
Be aware, that any solution should resolve (or at least allow further enhancement for) the problem of [running two cleaner processes concurrently](https://github.com/sosreport/sos/issues/3806).

So far we have these options:
- Jake's cleaner thread (that does not fix the #3806)
- Pavel's hashing approach (completely orthogonal to anything else, but with own gotchas, esp. risk of guessing whole mapping that is deterministically computable anywhere (with right salt))
- Nikhil's idea of interactivelly writing the mapping with locks (https://github.com/sosreport/sos/issues/3806#issuecomment-2457345169)

I didnt feel confident with either option, but after very fruitful discussions today, here are three more ideas that I will describe in detail in separate comments:
- have mapper process instead of Jake's thread where cleaners communicate with it via Unix file sockets
- cleaner processes/threads "compete" for the just-next hashed value by moving files into the right filename
- dont reinvent the wheel and use SQLite DB for the "tell me value or assign new one" transaction

Keep in mind that I think *NO* solution covers the use case "two concurrent cleaner processes with different and incompatible mapper files in input".

**Comment by pmoravec** [comment]
DAEMON-LIKE LISTENING ON SOCKET:
Have a dedicated mapper process for the Jake's idea:
- cleaner checks if a socket is open; if not it will spawn a mapper process that runs as a server for cleaner clients
- any client first advertise itself to the mapper process
- any cleaner / client keeps actual mapping in memory; for new strings to obfuscate, they query the mapper server via the socket, it replies back with a newly assigned obfuscated value
- once a cleaner is done, it requests the mapper process to write down the json file and announces its shutdown
- to prevent ghost mapper process after all cleaners are done: the mapper process keeps counter of running clients (or list of their PIDS) - once it reaches zero, mapper process shutdowns; for killed cleaner clients that didnt announce their shutdown, we can add a periodic probe of process PID liveness to prevent ghost mapper process

**Reply by TrevorBenson** [reply]
This idea sounds like it could easily lend itself to a distributed cleaning logic. At least when considering just the control persist I could see mapping the socket through ssh to each node, and then if the clean was passed to the nodes vs. pulled back to the collector the collector could provide the socket to a distributed cleaning.

I haven't spend much time thinking through any complications this might present, or how it would be solved with alternative transports. I'm not suggesting this be implemented in a way that only works with one transport, it's just a "off the top of my head" comment about this method and one transport in relation to the distributed cleaning topic.

**Comment by pmoravec** [comment]
FILE-BASED APPROACH (kudos to @adamruzicka for the idea):
Have a file-based cache where cleaners compete for new numbers for the mapping via atomic `mv` operation.
- the cache contains one directory per mapper class (hostname/IP/MAC/username/..)
- each directory will gradually contain files like host007 with content of the original hostname
- a cleaner wanting to obfuscate a hostname will:
  - read all unknown files if they already has that content - if so, the filename is the obfuscated value
  - if no file content matches, a new record must be created; the process/thread will create a temporary file with the hostname as the content
  - it will try to `mv` the file to the first not-existing filename
  - `mv` succeeded? then the process "won" and assigned that obfuscated value
  - if `mv` failed, some other process allocated the spot; go to beginning (read all unknown files)
- we might need up to 100k-ish files; this usually wont be a problem as most of FSes are on XFS that should have sufficient amount of (free?) inodes (but there will be corner cases like customer problem with inodes almost deplenished)
- writing final json file needs some synchronization, but can be also done in a safe way; the same applies to cleaning the cache

**Comment by pmoravec** [comment]
SQLITE (kudos to @bmr-cymru for the idea):
use `sqlite3` library; sos cleaners connect to a local db with tables for each mapper class, and leave the concurrent race of "the next obfuscated value" to the ACID feature of the DB (https://www.sqlite.org/transactional.html). They generate user-friendly mapper json file like today at the end, well, like today.

**Reply by TrevorBenson** [reply]
Would this method potentially result in having different values for the same obfuscated value from nodes in `sos collect --clean`? For example, if `/etc/hosts` had different orders of `IP Hostname(s)`, or other reasons, like different files containing the same value? 

Trying to determine if this provides an alternative to the hashing solution (in regards to unified obfuscation maps for sos collection with cleaning) or if it allows a distributed cleaning but could still result in unique obfuscation maps per node.

Thanks

**Reply by pmoravec** [reply]
The behaviour would be the same like today: recall that collector first grabs sosreports from each node, and then calls cleaner on each of them "locally". So all hostnames in the sosreports in the collector tarball will be obfuscated the same way. But if there would be a discrepancy between the mapping JSON files on individual nodes and the "main" node (where you call the collector from), then the same hostname can be obfuscated differently in sos collector tarball and in sosreports generated and cleaned locally on the nodes.

The same problem happens when you run `sos report --clean` on two systems (within the same cluster) independently - the mapping can easily be different until you ensure some synchronization of the mapping file among the systems (on your own).

And the same problem arises also when using hashing approach with salt - you have to copy the salt cluster-wide first, to ensure all mappings will be the same (but yes, you can do such sync just once, not everytime after each mapping file update).

The only solution how to prevent inconsistent mapping on different systems is using hashing approach without salt. But I feel that would put the whole obfuscation into too big risk: running a dictionary-like attack and using the known deterministic hashing algorithm, one can find the original hostname/IP/username/.. after some time (until the hashed value used in obfuscated string is sufficiently long (aka we use strong-enough hashing algorithm that produces such long hashes) to make such dict attack very time-lengthy. But then these hashes are user-unfriendly; would you like to see logs full of obfuscated strings like `hostname-5dd4df7d-e2d3-4d96-b512-2dc301fae3cc`? Recall that even `sha1sum` produces longer hexstreams as hashes than `uuidgen` in above example, while SHA-1 should not be used for security related purposes (I hope I am not mistaken here, please correct me if I am wrong wrt length of SHA1 hash values).

**Reply by TrevorBenson** [reply]
So all three new options are not actually "options for a distributed cleaning" (discussion primary topic), just new cleaning options persisting the non distributed cleaning methods?

EDIT: I don't consider this bad, just making sure if they are distributed or non distributed based on the initial topic and if any of them are (or do support) distributed cleaning.

**Reply by TrevorBenson** [reply]
> would you like to see logs full of obfuscated strings like `hostname-5dd4df7d-e2d3-4d96-b512-2dc301fae3cc`

Nope, preferably not.

My questions are mostly asked because of the initial topic of the thread, not because I find distributed as the only acceptable solution to discuss.

FWIW, the original topic came from having multiple environments where the collector is drastically under powered vs. the nodes it collects from. For example instances of the collector being a VM, having only 1-2 cores, only 4-8GB of RAM, or a limited amount of available filesystem space. Where each cluster member has at least 24 cores, 256GB of RAM, and 100GB+ more free space in `/var/`, and potentially has many different log files where using a time range based log collection requires increasing the 25MB limit, and could result in GB's of logs per node.

These are the things I have in "back of mind" as I ask additional questions, as a cleaning is critical for some of my use cases but somewhat impractical given the scale of clusters and resources on the collector. I also don't expect every idea has to address these shortcomings of a collector.

Thanks

**Reply by pmoravec** [reply]
> So all three new options are not actually "options for a distributed cleaning" (discussion primary topic), just new cleaning options persisting the non distributed cleaning methods?

Ah, I agree. I got confused / swapped two issues, this one and #3097 one, sorry for that hijaking of the topic. I will continue in the original one.

Definitely good questions to the individual ideas presented - I had similar ones that led e.g. to my gut feeling the hashing is cool orthogonal approach (that could work well in distributed environment) BUT I would be afraid of its bullet-proof security.

One potential distributed environment approach would be the "daemon-like listening on socket" - but listening on TCP socket, where nodes themselves run the cleaner and just query the collector node for new obfuscated strings of newly discovered sensitive strings. (and rather than implementing a new client-server solution, we can utilize the ssh/salt/.. connection from collector to the nodes for such communication..?)

**Reply by TrevorBenson** [reply]
Yep, I was thinking TCP for the socket one as an option vs. forwarding the unix socket mentioned in [this comment](https://github.com/orgs/sosreport/discussions/3476#discussioncomment-11788792). High security environments where all undefined ports are blocked at a firewall would be a limiting factor.

Although, once the TCP socket was documented the high security environments could simply use a change control request/window to open up the TCP connection from nodes back to to the expected collector if they wanted to have sos reports collected AND have a distributed cleaning with a unified obfuscation map, and then it would support any transport protocol.

**Comment by pmoravec** [comment]
Checking some old discussions, I think this can be closed. New concurrent cleaner has been implemented year-ish ago, now some work has been done in improving cleaner's performance (partly related topic).

I am closing the discussion for now. If somebody feels something is pending, let reopen it.
